### ISSUE TITLE
convert_image for YUV420p

### ISSUE THREAD

**Issue by HDaoud** [edrosten/libcvd#18]
Hello

I have a question regarding the conversion from YUV420p, in the video frame there is struct for 420p however, in the conversion it only accounts for 422, why is that?

how can I convert the yuv420p to rgb or bw?

Thanks

**Comment by edrosten** [comment]
On 21 May 2016 at 20:08, HDaoud <notifications@github.com> wrote:

> Hello
>
> I have a question regarding the conversion from YUV420p, in the video
> frame there is struct for 420p however, in the conversion it only accounts
> for 422, why is that?
>
> how can I convert the yuv420p to rgb or bw?
>
The YUV420 conversion code got removed temporarily because the old code had
no way to specify a stride which wasn't the same as the image size, and I
hadn't put in the mechanism to prevent special image types from having
nontrivial strides. I've not had a chance to put it back yet because I
don't have any data to actually test my code on (nothing I have produces
YUV420p data).

If you can send me some YUV420p data, then I can re-instate the code.

Basically, take the chunk of bytes and dump it to a file and send me the
file, along with the size of the image.


-Ed

**Comment by HDaoud** [comment]
Dear Edward,

Kindly find a small video of yuv420p, it's 640x368, 30Hz.
https://drive.google.com/open?id=0Bxnz_Y_iV3T1S3BIVnR1MG9yRnM

The extension of "yuv" was added manually to be played using vooya on linux.

Thank you!

**Comment by edrosten** [comment]
I've just pushed it to master.

I can now load bytes from the file you sent me, wrap it with a
BasicImage<yuv420p>, then convert_image it to an Image<Rgb<byte>> and save
it.

It looks about right

**Comment by edrosten** [comment]
Reopen if there are problems.

**Comment by HDaoud** [comment]
Dear Edward,

Thank you for the update.

I downloaded the new files and tried to recompile however, I'm getting bunch of error, kindly find the file attached.

[Error.txt](https://github.com/edrosten/libcvd/files/280115/Error.txt)

I checked the changes and in image.h there is only one function added for stride length. why all these errors?

I was searching online, and some suggesting it's due to the version of gcc, mine is 4.8.4

Kindly advise!

Thanks

**Comment by HDaoud** [comment]
I found an issue already opened with same errors, I'm now installing the gcc 5, will test and update accordingly.

thank you!

**Comment by HDaoud** [comment]
Hello Dear Edward,

After installing the gcc5 I managed to compile libcvd, and when trying to compile PTAMM I got errors that were fixed after adding "-std=c++14" to Flags in makefile. after compiling an error showed and it's due to the usage of unidentified function "totalsize()" in utility.h line#40 after changing that to in.size() another error shows :

`/usr/local/include/cvd/utility.h:40:41: error: no matching function for call to ‘CVD::Pixel::ConvertPixels<unsigned char, unsigned char, CVD::Pixel::GenericConversion<unsigned char, unsigned char, 1, 1>, true>::convert(CVD::Internal::ImageData<unsigned char, false>::ConstPointerType, CVD::Internal::ImageData<unsigned char, false>::PointerType, CVD::ImageRef)’
       Pixel::ConvertPixels<S,T>::convert(in.data(), out.data(), in.size());`

Any idea on that?

Should I post it as new issue?

Thank you for your patience,

**Comment by HDaoud** [comment]
Did you mean by `totalsize()` to return `size().x * size().y`?

as the error seems to be due to the compiler can't find the conversion between `size_t` and `ImageRef`

by the way as you closed this issue, I can't reopen it.

**Comment by HDaoud** [comment]
Dear Edward

Just to update you, I have changed the `totalsize()` to `size().x * size().y` according to my understanding of the code, please confirm if this is right.

and after compiling another error arose which is `error: call of overloaded ‘isnan(TooN::Internal::MultiplyType<double, double, 1>::type)’ is ambiguous` I have fixed that by adding the `std::` prior to `isnan(...)`, later another error showed which is `PatchFinder.cc:112:18: error: ‘transform’ is not a member of ‘CVD’` I tried to reconfigure and install `libcvd` but that didn't help, then I deleted all files in /usr/local/include and `make clean` all projects, and re-installed TooN, libcvd, gvars3 and lib3ds however, the error is still showing! I'm stuck here now and appreciate your help.

Regards,

**Comment by edrosten** [comment]
Hi,

So, totalsize() became .size().area(). The main reason for .totalsize() being removed was that it was almost always used in things iterating over pixels under the assumption that they were contiguous. Since the case is now that all normal images (this excludes funny types like YUV422, 420p, bayer, etc) can now have an arbitrary stride between rows, almost every use of totalsize() was incorrect.  Contiguous images used to be the only option because there used to be a small win, but optimizers have got better, so the win more or less disappeared. 

I've pushed a speculative fix for this compile error. Can you check to see if it works? utility.h is old, hoary code and isn't covered by any of the (rather sparse) unit tests for the time being.

To be honest, the whole copy() function is now redundant given that sub_image() and convert_image exist. I won't delete it though of course!

Is the isnan thing a fix for TooN? Also, are you on a TooN2 release or the master branch?

There's a transform function in cvd/vision.h. It may be that the include files have changed slightly so it's not being pulled in where it was before. Likewise there's a std::transform in <algorithm>. Do either of those do what's necessary?

-Ed

**Comment by HDaoud** [comment]
Hello Edward,

I was using TooN2 but to make sure that my code similar to what expected is yours, I downloaded the TooN-master and now I'm using it.

In my trial to fix the `transform` error, I had to comment out the `#if defined (CVD_HAVE_TOON)` I don't know why compiler may failing here however, it fixed the error.

all files are compiled, and objects are generated but for some reason I'm getting bunch of undefined reference, kindly find them below.

`System.o: In function `PTAMM::System::StartMapSerialization(std::string, std::string)':
System.cc:(.text+0x7f0): undefined reference to `CVD::Thread::start(CVD::Runnable*)'
MapMaker.o: In function `PTAMM::MapMaker::~MapMaker()':
MapMaker.cc:(.text+0x99): undefined reference to `CVD::Thread::stop()'
MapMaker.cc:(.text+0xf0): undefined reference to `CVD::Thread::join()'
MapMaker.cc:(.text+0x1d5): undefined reference to `CVD::Thread::~Thread()'
MapMaker.o: In function `PTAMM::MapMaker::MapMaker(std::vector<PTAMM::Map*, std::allocator<PTAMM::Map*> >&, PTAMM::Map*)':
MapMaker.cc:(.text+0x7bb8): undefined reference to `CVD::Thread::Thread()'
MapMaker.cc:(.text+0x7c46): undefined reference to `CVD::Thread::start(CVD::Runnable*)'
MapMaker.cc:(.text+0x7e8a): undefined reference to `CVD::Thread::~Thread()'
MapMaker.o: In function `PTAMM::MapMaker::run()':
MapMaker.cc:(.text+0x81e7): undefined reference to `CVD::Thread::shouldStop() const'
MapMaker.cc:(.text+0x8258): undefined reference to `CVD::Thread::shouldStop() const'
MapMaker.cc:(.text+0x82ad): undefined reference to `CVD::Thread::sleep(unsigned int)'
MapMaker.o:(.rodata._ZTIN5PTAMM8MapMakerE[_ZTIN5PTAMM8MapMakerE]+0x18): undefined reference to `typeinfo for CVD::Thread'
MapSerializer.o: In function `PTAMM::MapSerializer::MapSerializer(std::vector<PTAMM::Map*, std::allocator<PTAMM::Map*> >&)':
MapSerializer.cc:(.text+0x8f): undefined reference to `CVD::Thread::Thread()'
MapSerializer.o: In function `PTAMM::MapSerializer::Init(std::string, std::string, PTAMM::Map&)':
MapSerializer.cc:(.text+0x261): undefined reference to `CVD::Thread::isRunning() const'
MapSerializer.o: In function `PTAMM::MapSerializer::~MapSerializer()':
MapSerializer.cc:(.text+0xe5d): undefined reference to `CVD::Thread::~Thread()'
MapSerializer.cc:(.text+0xe93): undefined reference to `CVD::Thread::~Thread()'
MapSerializer.o:(.rodata._ZTIN5PTAMM13MapSerializerE[_ZTIN5PTAMM13MapSerializerE]+0x10): undefined reference to `typeinfo for CVD::Thread'
Model3ds.o: In function `PTAMM::Model3DS::_GenerateDisplayList(Lib3dsFile*, bool)':
Model3ds.cc:(.text+0x2bf): undefined reference to `lib3ds_mesh_calculate_vertex_normals'
Model3ds.o: In function `PTAMM::Model3DS::_Load()':
Model3ds.cc:(.text+0x8db): undefined reference to `lib3ds_file_open'
collect2: error: ld returned 1 exit status
make: *** [PTAMM] Error 1`

**Comment by edrosten** [comment]
On 25 May 2016 at 13:30, HDaoud <notifications@github.com> wrote:

> Hello Edward,
>
> I was using TooN2 but to make sure that my code similar to what expected
> is yours, I downloaded the TooN-master and now I'm using it.
>

It shouldn't matter either way.


> In my trial to fix the transform error, I had to comment out the #if
> defined (CVD_HAVE_TOON) I don't know why compiler may failing here
> however, it fixed the error.
>

Seems like the TooN detection is failing somehow. I'm not sure why it
would, I'll look.


> all files are compiled, and objects are generated but for some reason I'm
> getting bunch of undefined reference, kindly find them below.
>
> System.o: In functionPTAMM::System::StartMapSerialization(std::string,
> std::string)':
> System.cc:(.text+0x7f0): undefined reference to
> CVD::Thread::start(CVD::Runnable*)'
> MapMaker.o: In functionPTAMM::MapMaker::~MapMaker()':
> MapMaker.cc:(.text+0x99): undefined reference to CVD::Thread::stop()'
> MapMaker.cc:(.text+0xf0): undefined reference toCVD::Thread::join()'
> MapMaker.cc:(.text+0x1d5): undefined reference to CVD::Thread::~Thread()'
> MapMaker.o: In functionPTAMM::MapMaker::MapMaker(std::vector >&,
> PTAMM::Map
>
> *)': MapMaker.cc:(.text+0x7bb8): undefined reference to
> CVD::Thread::Thread()' MapMaker.cc:(.text+0x7c46): undefined reference
> toCVD::Thread::start(CVD::Runnable*)'
> MapMaker.cc:(.text+0x7e8a): undefined reference to CVD::Thread::~Thread()'
> MapMaker.o: In functionPTAMM::MapMaker::run()':
> MapMaker.cc:(.text+0x81e7): undefined reference to CVD::Thread::shouldStop()
> const'
> MapMaker.cc:(.text+0x8258): undefined reference toCVD::Thread::shouldStop()
> const'
> MapMaker.cc:(.text+0x82ad): undefined reference to CVD::Thread::sleep(unsigned
> int)'
> MapMaker.o:(.rodata._ZTIN5PTAMM8MapMakerE[_ZTIN5PTAMM8MapMakerE]+0x18):
> undefined reference totypeinfo for CVD::Thread'
> MapSerializer.o: In function PTAMM::MapSerializer::MapSerializer(std::vector<PTAMM::Map*,
> std::allocator<PTAMM::Map*> >&)':
> MapSerializer.cc:(.text+0x8f): undefined reference to
> CVD::Thread::Thread()'
> MapSerializer.o: In function PTAMM::MapSerializer::Init(std::string,
> std::string, PTAMM::Map&)':
> MapSerializer.cc:(.text+0x261): undefined reference toCVD::Thread::isRunning()
> const'
> MapSerializer.o: In function PTAMM::MapSerializer::~MapSerializer()':
> MapSerializer.cc:(.text+0xe5d): undefined reference to
> CVD::Thread::~Thread()'
> MapSerializer.cc:(.text+0xe93): undefined reference to
> CVD::Thread::~Thread()'
> MapSerializer.o:(.rodata._ZTIN5PTAMM13MapSerializerE[_ZTIN5PTAMM13MapSerializerE]+0x10):
> undefined reference totypeinfo for CVD::Thread'
> Model3ds.o: In function PTAMM::Model3DS::_GenerateDisplayList(Lib3dsFile*,
> bool)':
> Model3ds.cc:(.text+0x2bf): undefined reference to
> lib3ds_mesh_calculate_vertex_normals'
> Model3ds.o: In function PTAMM::Model3DS::_Load()':
> Model3ds.cc:(.text+0x8db): undefined reference tolib3ds_file_open'
> collect2: error: ld returned 1 exit status
> make: *** [PTAMM] Error 1`
>

Ah. Well, that's annoying!

OK firstly, you've probably got an old version of the CVD headers
installed: CVD::thread was removed, so it shouldn't have compiled. Likely
you have a the thread header kicking around.

Secondly, it's annoying because being based on C++98 originally, libCVD
sprouted all sorts of dubiously well maintained platform compatibility
stuff to make porting easily. I stripped out everything that duplicated
C++14 features, since those are now idiomatic, better maintained and often
more nicely designed!

It seems like you've been bitten by me removing CVD::thread.

Given that code out there in the wild uses it, I think I should reinstate
it, but using std::thread instead of pthreads.

-Ed

**Comment by HDaoud** [comment]
I think that's because I copied the new files to the old folder (merged and replaced) so the files that is deleted in the new package is still showing from the old one :sweat_smile: 

What can I do to fix it?

Thanks

**Comment by edrosten** [comment]
I've reinstated CVD::thread (with lots of deprecated warnings), but it's
lightly tested, so there may be bugs to fix.

**Comment by HDaoud** [comment]
Ok, I will try and let you know, btw when using `conver_image` instead of `total_length` it gives `error: ‘convert_image’ was not declared in this scope``
       convert_image(in, out);`

**Comment by HDaoud** [comment]
actually I just added `cvd_src/thread.o` to libcvd makefile and it seems to solve the previous undefined errors, as I still have the thread.h and thread.cpp however, it show error in lib3ds:

`Model3ds.o: In function `PTAMM::Model3DS::_GenerateDisplayList(Lib3dsFile*, bool)':
Model3ds.cc:(.text+0x2bf): undefined reference to `lib3ds_mesh_calculate_vertex_normals'
Model3ds.o: In function `PTAMM::Model3DS::_Load()':
Model3ds.cc:(.text+0x8db): undefined reference to `lib3ds_file_open'
collect2: error: ld returned 1 exit status
make: *** [PTAMM] Error 1`

**Comment by HDaoud** [comment]
Model3DS errors were fixed after downgrading from 1.3 to the 20080909.

Thanks for your help Edward, I'm closing this issue.

**Comment by edrosten** [comment]
Glad that worked. Note, after the recent push, because Makefile.in changed, you'd need to rerun config.status (or just rerun ./configure), in order to generate a new Makefile.

**Comment by TengFeiHan0** [comment]
@HDaoud, I also got the same error when running libcvd,
`libcvd/include/cvd/utility.h:40:17: error: ‘convert_image’ was not declared in this scope
    convert_image(in, out);
could you tell me how did you had fixed it? thank you in advance

**Comment by edrosten** [comment]
Hi @TengFeiHan0,

Can you post a small program to the thread that reveals this error?

Also, can you let me know which library version (e.g. which git hash you're
on) and which compiler you're using?

-Ed
